In [1]:
# Setup

import pandas as pd
import numpy as np
from sklearn.tree import (
    DecisionTreeClassifier,
    DecisionTreeRegressor,
    export_text
)
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, mean_absolute_error
from sklearn.ensemble import RandomForestClassifier
from sklearn.ensemble import RandomForestRegressor
from xgboost import XGBClassifier, XGBRegressor

df = pd.read_csv('heart_disease.csv')

In [2]:
df

,age,sex,cp,trestbps,chol,fbs,restecg,thalach,exang,oldpeak,slope,ca,thal,target
0,70,1,3,130,322,0,2,109,0,2.4,1,3,1,1
1,67,0,2,115,564,0,2,160,0,1.6,1,0,3,0
2,57,1,1,124,261,0,0,141,0,0.3,0,0,3,1
3,64,1,3,128,263,0,0,105,1,0.2,1,1,3,0
4,74,0,1,120,269,0,2,121,1,0.2,0,1,1,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
265,52,1,2,172,199,1,0,162,0,0.5,0,0,3,0
266,44,1,1,120,263,0,0,173,0,0.0,0,0,3,0
267,56,0,1,140,294,0,2,153,0,1.3,1,0,1,0
268,57,1,3,140,192,0,0,148,0,0.4,1,0,2,0


In [3]:
# Task 1

X = df.drop('target', axis=1)
y = df['target']

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)

print(X_train.shape)
print(X_test.shape)

(216, 13)
(54, 13)


In [4]:
# Task 2

dt = DecisionTreeClassifier(random_state=42)
dt.fit(X_train, y_train)

print(f"Training Accuracy: {accuracy_score(y_train, dt.predict(X_train)):.2}")
print(f"Testing Accuracy: {accuracy_score(y_test, dt.predict(X_test)):.2}")

# The model is overfit because the training accuracy is much higher than the test

Training Accuracy: 1.0
Testing Accuracy: 0.69


In [5]:
# Task 3

print(export_text(dt, feature_names=X.columns))

num_leaves = dt.tree_.n_leaves
print(f"Number of leaves: {num_leaves}")

depth = dt.get_depth()
print(f"Depth: {depth}")

|--- cp <= 2.50
|   |--- oldpeak <= 2.45
|   |   |--- thal <= 1.50
|   |   |   |--- chol <= 272.00
|   |   |   |   |--- trestbps <= 109.00
|   |   |   |   |   |--- chol <= 232.50
|   |   |   |   |   |   |--- class: 0
|   |   |   |   |   |--- chol >  232.50
|   |   |   |   |   |   |--- class: 1
|   |   |   |   |--- trestbps >  109.00
|   |   |   |   |   |--- class: 0
|   |   |   |--- chol >  272.00
|   |   |   |   |--- oldpeak <= 1.35
|   |   |   |   |   |--- thalach <= 138.00
|   |   |   |   |   |   |--- class: 1
|   |   |   |   |   |--- thalach >  138.00
|   |   |   |   |   |   |--- chol <= 330.00
|   |   |   |   |   |   |   |--- class: 0
|   |   |   |   |   |   |--- chol >  330.00
|   |   |   |   |   |   |   |--- chol <= 337.50
|   |   |   |   |   |   |   |   |--- class: 1
|   |   |   |   |   |   |   |--- chol >  337.50
|   |   |   |   |   |   |   |   |--- class: 0
|   |   |   |   |--- oldpeak >  1.35
|   |   |   |   |   |--- class: 1
|   |   |--- thal >  1.50
|   |   |   |--- thalac

1. Depth is 10
2. 36 leaves
3. Is cp less than or equal to 2.50?

In [7]:
# Task 4
rf = RandomForestClassifier(n_estimators=100, random_state=42)
rf.fit(X_train, y_train)

print(f"Training Accuracy: {accuracy_score(y_train, rf.predict(X_train)):.2}")
print(f"Testing Accuracy: {accuracy_score(y_test, rf.predict(X_test)):.2}")

Training Accuracy: 1.0
Testing Accuracy: 0.8


The test accuracy improved from using RandomForest. The gap got smaller. The DecisionTree over fit more.

In [9]:
#Task 5
xgb = XGBClassifier(n_estimators=100, learning_rate=0.1, random_state=42, eval_metric='logloss')
xgb.fit(X_train, y_train)

print(f"Training Accuracy: {accuracy_score(y_train, xgb.predict(X_train)):.2}")
print(f"Testing Accuracy: {accuracy_score(y_test, xgb.predict(X_test)):.2}")

Training Accuracy: 1.0
Testing Accuracy: 0.85


The XGBoost test did slightly better than RandomForest.

XGBoost closed the gap but only slightly.

|Model       | Train Accuracy | Test Accuracy  |
|:------------:|------------:|
|Random Forest|        |        100  |
| XGBoost     |    Elephant  |         20  |